breakerbp/failure_ratio.go

```diff
@@ -0,0 +1,131 @@
+package breakerbp
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/apache/thrift/lib/go/thrift"
+	"github.com/sony/gobreaker"
+
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+)
+
+// FailureRatioBreaker is a circuit breaker based on gobreaker that uses a low-water-mark and
+// % failure threshold to trip.
+type FailureRatioBreaker struct {
+	goBreaker *gobreaker.CircuitBreaker
+
+	name              string
+	minRequestsToTrip int
+	failureThreshold  float64
+	logger            log.Wrapper
+}
+
+// Config represents the configuration for a FailureRatioBreaker.
+type Config struct {
+	// Minimum requests that need to be sent during a time period before the breaker is eligible to transition from closed to open.
+	MinRequestsToTrip int
+
+	// Percentage of requests that need to fail during a time period for the breaker to transition from closed to open.
+	// Represented as a float in [0,1], where .05 means >=5% failures will trip the breaker.
+	FailureThreshold float64
+
+	// Name for this circuit breaker, mostly used as a prefix to disambiguate logs when multiple cb are used.
+	Name string
+
+	// EmitStatusMetrics sets whether the failure breaker will regularly update a gauge on the breakers state (closed or open/halfopen)
+	// When enabled, it emits metrics using the interval defined by metricsbp.SysStatsTickerInterval.
+	EmitStatusMetrics bool
+
+	// Logger is the logger to be called when the breaker changes states.
+	Logger log.Wrapper
+
+	// MaxRequestsHalfOpen represents he Maximum amount of requests that will be allowed through while the breaker
+	// is in half-open state. If left unset (or set to 0), exactly 1 request will be allwed through while half-open.
+	MaxRequestsHalfOpen uint32
+
+	// Interval represents the cyclical period of the 'Closed' state.
+	// If 0, internal counts do not get reset while the breaker remains in the Closed state.
+	Interval time.Duration
+
+	// Timeout is the duration of the 'Open' state. After an 'Open' timeout duration has passed, the breaker enters 'half-open' state.
+	Timeout time.Duration
+}
+
+// NewFailureRatioBreaker creates a new FailureRatioBreaker with the provided configuration. Creates a new goroutine to emit
+// breaker state metrics if EmitStatusMetrics is set to true.
```

Comment:
It would be good to note in the comment when the goroutine is stopped (when `metricsbp.M.Ctx()` is `Done()`).